core/src/main/java/com/linecorp/armeria/client/endpoint/SlowStartAwareEndpointSelectionStrategy.java

```diff
@@ -0,0 +1,175 @@
+    private final ScheduledExecutorService executorService;
+    private final Duration slowStartInterval;
+    private final int numberOfSteps;
+
+    /**
+     * Creates a {@link SlowStartAwareEndpointSelectionStrategy}.
+     */
+    SlowStartAwareEndpointSelectionStrategy(EndpointWeightTransition endpointWeightTransition,
+                                            ScheduledExecutorService executorService,
+                                            Duration slowStartInterval,
+                                            int numberOfSteps) {
+        this.endpointWeightTransition = endpointWeightTransition;
+        this.executorService = executorService;
+        this.slowStartInterval = requireNonNull(slowStartInterval, "slowStartInterval");
+        checkArgument(numberOfSteps > 0, "numberOfSteps(%s) > 0");
+        this.numberOfSteps = numberOfSteps;
+    }
+
+    @Override
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new SlowStartAwareWeightedRoundRobinSelector(endpointGroup, endpointWeightTransition);
+    }
+
+    /**
+     * An {@link EndpointSelector} that is similar to {@link WeightedRoundRobinStrategy} but it
+     * increases each {@link Endpoint} weight gradually.
+     */
+    private final class SlowStartAwareWeightedRoundRobinSelector implements EndpointSelector {
+        private final EndpointGroup endpointGroup;
+        private final AtomicInteger sequence = new AtomicInteger();
+        private volatile EndpointsAndWeights endpointsAndWeights;
+        private final EndpointWeightTransition endpointWeightTransition;
+
+        private volatile ScheduledFuture<?> updateEndpointWeightFuture;
+        private final Map<Endpoint, EndpointAndStep> currentWeights = new ConcurrentHashMap<>();
+
+        SlowStartAwareWeightedRoundRobinSelector(EndpointGroup endpointGroup,
+                                                 EndpointWeightTransition endpointWeightTransition) {
+            this.endpointGroup = endpointGroup;
+            this.endpointWeightTransition = endpointWeightTransition;
+            final List<Endpoint> endpoints = endpointGroup.endpoints();
+            endpointsAndWeights = new EndpointsAndWeights(endpoints);
+            endpointGroup.addListener(this::scheduleUpdateRequestWeight);
+        }
+
+        @Override
+        public EndpointGroup group() {
+            return endpointGroup;
+        }
+
+        @Override
+        public EndpointSelectionStrategy strategy() {
+            return SlowStartAwareEndpointSelectionStrategy.this;
+        }
+
+        @Override
+        public Endpoint select(ClientRequestContext ctx) {
+            final int currentSequence = sequence.getAndIncrement();
+            return endpointsAndWeights.selectEndpoint(currentSequence);
+        }
+
+        private void scheduleUpdateRequestWeight(List<Endpoint> newEndpoints) {
+            if (updateEndpointWeightFuture != null && !updateEndpointWeightFuture.cancel(false)) {
```

Comment:
I think this method can be called concurrently, since `addListener` notifications can happen in quick succession and don't have any fixed thread model IIUC.

Simplest solution is probably to make this method `synchronized`.